### PR TITLE
8317545: AIX PPC64: Implementation of Foreign Function & Memory API

### PR DESCRIPTION
**Pull request by TheRealMDoerr**
The AIX linker has a few minor diffs to the linux ABIv1 linker. In addition, double values have only 4 Byte alignment.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317545](https://bugs.openjdk.org/browse/JDK-8317545): AIX PPC64: Implementation of Foreign Function &amp; Memory API (**Enhancement** - P3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16043/head:pull/16043` \
`$ git checkout pull/16043`

Update a local copy of the PR: \
`$ git checkout pull/16043` \
`$ git pull https://git.openjdk.org/jdk.git pull/16043/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16043`

View PR using the GUI difftool: \
`$ git pr show -t 16043`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16043.diff">https://git.openjdk.org/jdk/pull/16043.diff</a>

</details>

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
@JornVernee: Doubles use 4 Byte alignment on AIX. Please take a look.
Many tests are not passing because symbol lookup will need an enhancement. Suchismith Roy from IBM is working on that. (Should probably be done as a subtask.)

File: src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java [hunk 1]
```diff
@@ -283,6 +284,7 @@ private static FunctionDescriptor stripNames(FunctionDescriptor function) {
             ValueLayout.JAVA_FLOAT,
             ValueLayout.JAVA_LONG,
             ValueLayout.JAVA_DOUBLE,
-            ValueLayout.ADDRESS
+            ValueLayout.ADDRESS,
+            ValueLayout.JAVA_DOUBLE.withByteAlignment(4) // used by AIX
```
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:288]: We can't add the layout here. The supported layout check essentially tries to verify that the layout used for e.g. `double`, matches the layout of the `C` double on that platform. So, double with 4 byte alignment should not be allowed on other platforms.

I think we should add a linker-specific supported layouts check instead.

I suggest making `checkSupported` non-static, and then adding an abstract method to check if a layout is supported that is called from `checkSupported`. See for instance how it's done for the `cannonicalLayouts` in the panama-foreign repo: https://github.com/openjdk/panama-foreign/blob/foreign-memaccess%2Babi/src/java.base/share/classes/jdk/internal/foreign/abi/x64/windows/Windowsx64Linker.java#L44-L45 this uses a shared factory method in SharedUtils to create the Map, where each linker impl can specify some platform specific layouts, and have it's own instance of the Map. Something similar should work in this case I think.

Comment (TheRealMDoerr) [reply src/java.base/share/classes/jdk/internal/foreign/abi/AbstractLinker.java:288]: Thanks for the quick response! We'll try (maybe after getting more tests working, there could be more).

**Comment by JornVernee** [comment]
Actually, looking at the latest code again we have in the JEP: we define the supported layouts as a set of all the canonical layouts, which is also what the spec mandates in the latest version. (See how this is implemented in the JEP, in `AbstractLinker`: https://github.com/openjdk/jdk/pull/15103/files#diff-6b6ffc0ec79ecb2e54ae2a42c9f082db1fb7d5d2a5e1a02b0b368274b7c27425R77). Sorry, I had forgotten that we'd made that change.

Given that we are getting closer to the JEP integration, it might be better to base the work of this patch on the changes in the latest JEP PR instead, since the AIX linker needs changes to the supported layouts as well. Also, the current specification in the `jdk` repo ([1]) doesn't leave any room for a 4-byte-aligned JAVA_DOUBLE layout to be a supported linker layout, so implementing the AIX linker ahead of the JEP would require a separate spec change/CSR. The latest spec as found in the JEP patch, on the other hand, _does_ allow for a 4-byte-aligned JAVA_DOUBLE (by saying that all canonical layouts are supported, and those are defined per platform).

So, it seems simpler to have this PR go in after the JEP.

[1]: https://github.com/openjdk/jdk/blob/a1c9587c27538bda3b0f6745d9c80ff4e1b9a77e/src/java.base/share/classes/java/lang/foreign/Linker.java#L204-L215

**Comment by TheRealMDoerr** [comment]
Thanks for the information. I can rebase and adapt once the JEP is integrated. That's fine with me. We need some more time anyways. We will need the symbol resolution working before we can really test it (Suchi's work).
I have created this PR because the fallback linker causes problems on AIX. Maybe it doesn't support int to long conversions as needed by the ABI.

File: src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/aix/AixPPC64Linker.java [hunk 2]
```diff
@@ -62,4 +65,22 @@ protected UpcallStubFactory arrangeUpcall(MethodType targetType, FunctionDescrip
...
+
+    protected static final Set<MemoryLayout> AIX_LAYOUTS = Set.of(
+            ValueLayout.JAVA_BOOLEAN,
+            ValueLayout.JAVA_BYTE,
+            ValueLayout.JAVA_CHAR,
+            ValueLayout.JAVA_SHORT,
+            ValueLayout.JAVA_INT,
+            ValueLayout.JAVA_FLOAT,
+            ValueLayout.JAVA_LONG,
+            ValueLayout.JAVA_DOUBLE,
```
[the remark this answers is not in the document]
Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/ppc64/aix/AixPPC64Linker.java:77]: Yes. The tests would likely need to be updated as well, as they define C_DOUBLE using JAVA_DOUBLE. The latest version (in the JEP) does it the right way though: it defines C_DOUBLE using Linker.canonicalLayouts.

We've also ensured that tests pass on x86_32/Linux ([1]), which also has 4 byte alignment for double. So, in general there should be less obstacles in the latest version of the code.

[1]: https://github.com/openjdk/panama-foreign/pull/849

**Comment by JornVernee** [comment]
> I have created this PR because the fallback linker causes problems on AIX. Maybe it doesn't support int to long conversions as needed by the ABI.

We did find some issues in libffi on Mac/AArch64: https://bugs.openjdk.org/browse/JDK-8309954

Is it possible that this is a bug in libffi itself? In principle the fallback linker is just a wrapper around libffi.

**Comment by TheRealMDoerr** [comment]
I didn't look into libffi / fallback linker. Suchi has observed problems. He is planning to check if this PR works better.

**Comment by TheRealMDoerr** [comment]
Closing in favor of https://github.com/openjdk/jdk/pull/16179.

**Pull request closed without merging**